File: examples/run_stereo.rs
use serde::{Deserialize, Serialize};
use std::{collections::HashMap, io::Cursor, path::PathBuf};

use anyhow::Ok;
use camera_intrinsic_model::io::model_from_json;
use clap::Parser;
use glob::glob;
use nalgebra as na;
use toy_stereo_vo::estimator::{Estimator, EstimatorParameters};

#[derive(Parser)]
#[command(version, about, long_about = None)]
struct StereoVoArgs {
    #[arg(short, long)]
    dataset_folder: String,

    #[arg(short, long)]
    config_folder: String,

    #[arg(long, action)]
    rerun: bool,
}

#[derive(Serialize, Deserialize, Clone, Copy)]
struct RvecTvec {
    pub rvec: [f32; 3],
    pub tvec: [f32; 3],
}
impl RvecTvec {
    pub fn to_isometry3(&self) -> na::Isometry3<f32> {
        na::Isometry3::new(
            na::Vector3::new(self.tvec[0], self.tvec[1], self.tvec[2]),
            na::Vector3::new(self.rvec[0], self.rvec[1], self.rvec[2]),
        )
    }
}

#[derive(Serialize, Deserialize, Clone)]
struct RvecTvecs {
    pub rtvecs: Vec<RvecTvec>,
}

fn rtvec_from_json(file_path: &str) -> RvecTvec {
    let contents =
        std::fs::read_to_string(file_path).expect("Should have been able to read the file");
    let rtvecs: RvecTvecs = serde_json::from_str(&contents).unwrap();
    rtvecs.rtvecs[1]
}

fn id_to_color(id: u64) -> [u8; 3] {
    const M: u32 = 2u32.pow(24);
    fastrand::seed(id);
    let color_num = fastrand::u32(0..M);
    [
        ((color_num >> 16) % 256) as u8,
        ((color_num >> 8) % 256) as u8,
        (color_num % 256) as u8,
    ]
}

fn load_camera_images(root: &str, cam_idx: usize) -> anyhow::Result<Vec<PathBuf>> {
    let pattern = PathBuf::from(root).join(format!("mav0/cam{}/data/*.png", cam_idx));

    glob(pattern.to_str().unwrap())?
        .map(|p| p.map_err(anyhow::Error::from))
        .collect()
}

fn log_rerun_image(rec: &rerun::RecordingStream, image: &image::GrayImage, entity_path: &str) {
    let mut bytes: Vec<u8> = Vec::new();
    image
        .write_to(&mut Cursor::new(&mut bytes), image::ImageFormat::Jpeg)
        .expect("fail to write jpeg");
    rec.log(entity_path, &rerun::EncodedImage::from_file_contents(bytes))
        .unwrap();
}

fn log_image_keypoints(
    rec: &rerun::RecordingStream,
    curr_keypoints: &HashMap<usize, (f32, f32)>,
    prev_keypoints: &HashMap<usize, (f32, f32)>,
    image_entity_path: &str,
) {
    // Cam0 points and lines
    let (colors0, points0): (Vec<_>, Vec<(f32, f32)>) = curr_keypoints
        .iter()
        .map(|(&id, &(x, y))| {
            let color = id_to_color(id as u64);
            (color, (x + 0.5, y + 0.5))
        })
        .unzip();
    rec.log(
        format!("{}/points", image_entity_path),
        &rerun::Points2D::new(points0).with_colors(colors0),
    )
    .unwrap();

    let mut line_strips0 = Vec::new();
    let mut line_colors0 = Vec::new();
    for (&id, &(curr_x, curr_y)) in curr_keypoints {
        if let Some(&(prev_x, prev_y)) = prev_keypoints.get(&id) {
            line_strips0.push([(prev_x + 0.5, prev_y + 0.5), (curr_x + 0.5, curr_y + 0.5)]);
            line_colors0.push(id_to_color(id as u64));
        }
    }
    if !line_strips0.is_empty() {
        rec.log(
            format!("{}/lines", image_entity_path),
            &rerun::LineStrips2D::new(line_strips0).with_colors(line_colors0),
        )
        .unwrap();
    }
}

fn to_rerun_transform(transform: &na::Isometry3<f32>) -> rerun::Transform3D {
    let qxyzw =
        na::UnitQuaternion::from_rotation_matrix(&transform.rotation.to_rotation_matrix()).coords;
    let rotation = rerun::Quaternion::from_xyzw([qxyzw[0], qxyzw[1], qxyzw[2], qxyzw[3]]);
    rerun::Transform3D::from_translation_rotation(
        [
            transform.translation.x,
            transform.translation.y,
            transform.translation.z,
        ],
        rotation,
    )
}

fn log_pose(
    rec: &rerun::RecordingStream,
    transform: &na::Isometry3<f32>,
    entity_path: &str,
    size: f32,
    option_camera_frustum: Option<rerun::Pinhole>,
) {
    rec.log(entity_path, &to_rerun_transform(transform))
        .unwrap();
    if let Some(camera_frustum) = option_camera_frustum {
        rec.log(entity_path, &camera_frustum).unwrap();
    } else {
        rec.log(entity_path, &rerun::TransformAxes3D::new(size))
            .unwrap();
    }
}

fn log_active_landmarks(
    rec: &rerun::RecordingStream,
    landmarks: &HashMap<usize, na::Point3<f32>>,
    current_t_w_cam0: &na::Isometry3<f32>,
    entity_path: &str,
    range_threshold: f32,
) {
    let range_threshold2 = range_threshold * range_threshold;
    let cam0_position = na::Point3::from(current_t_w_cam0.translation.vector);
    let (colors, points): (Vec<[u8; 3]>, Vec<[f32; 3]>) = landmarks
        .iter()
        .filter_map(|(&feature_id, &point)| {
            let distance2 = (point - cam0_position).norm_squared();
            if distance2 < range_threshold2 {
                Some((id_to_color(feature_id as u64), [point.x, point.y, point.z]))
            } else {
                None
            }
        })
        .unzip();
    rec.log(
        entity_path,
        &rerun::Points3D::new(points).with_colors(colors),
    )
    .unwrap();
}

fn log_old_landmarks(
    rec: &rerun::RecordingStream,
    landmarks: &[na::Point3<f32>],
    entity_path: &str,
) {
    if landmarks.is_empty() {
        return;
    }
    let points: Vec<[f32; 3]> = landmarks.iter().map(|pt| [pt.x, pt.y, pt.z]).collect();
    rec.log(entity_path, &rerun::Points3D::new(points)).unwrap();
}

fn log_trajectory(
    rec: &rerun::RecordingStream,
    trajectory: &[na::Isometry3<f32>],
    entity_path: &str,
) {
    let line_strips: Vec<[f32; 3]> = trajectory
        .iter()
        .map(|t_cam0_w| {
            let tvec_w = t_cam0_w.inverse().translation.vector;
            [tvec_w.x, tvec_w.y, tvec_w.z]
        })
        .collect();
    rec.log(entity_path, &rerun::LineStrips3D::new([line_strips]))
        .unwrap();
}

fn camera_model_to_rerun_pinhole(
    model: &camera_intrinsic_model::GenericModel<f64>,
) -> rerun::Pinhole {
    let projection_matrix = model
        .estimate_new_camera_matrix_for_undistort(0.0, None)
        .cast();
    rerun::Pinhole::from_focal_length_and_resolution(
        [projection_matrix[(0, 0)], projection_matrix[(1, 1)]],
        [model.width() as f32, model.height() as f32],
    )
    .with_principal_point([projection_matrix[(0, 2)], projection_matrix[(1, 2)]])
    .with_image_plane_distance(0.1)
}

fn main() -> anyhow::Result<()> {
    env_logger::init();

    let cli = StereoVoArgs::parse();

    let root_folder = cli.dataset_folder;

    let camera_images: Vec<Vec<PathBuf>> = (0..2)
        .map(|i| load_camera_images(&root_folder, i))
        .collect::<Result<_, _>>()?;

    anyhow::ensure!(
        camera_images[0].len() == camera_images[1].len() && !camera_images[0].is_empty(),
        "Mismatched or empty camera images"
    );
    println!("Found {} stereo image pairs", camera_images[0].len());

    // load camera models and extrinsics
    let config_folder = cli.config_folder;
    let left_cam_model = model_from_json(&format!("{}/cam0.json", config_folder));
    let right_cam_model = model_from_json(&format!("{}/cam1.json", config_folder));
    let t_cam1_cam0 = rtvec_from_json(&format!("{}/extrinsics.json", config_folder)).to_isometry3();

    let mut estimator = Estimator::new(
        left_cam_model.cast(),
        right_cam_model.cast(),
        t_cam1_cam0,
        Some(EstimatorParameters::default()),
    );
    let (rec, camera_frustum) = if cli.rerun {
        (
            Some(
                rerun::RecordingStreamBuilder::new("vo")
                    .spawn_opts(&rerun::SpawnOptions {
                        port: 9875,
                        ..Default::default()
                    })
                    .unwrap(),
            ),
            Some(camera_model_to_rerun_pinhole(&left_cam_model.cast())),
        )
    } else {
        (None, None)
    };

    if let Some(rec) = &rec {
        rec.log_static("/", &rerun::ViewCoordinates::RDF()).unwrap();
    }

    let mut prev_points0: HashMap<usize, (f32, f32)> = HashMap::new();
    let mut prev_points1: HashMap<usize, (f32, f32)> = HashMap::new();

    let start_time = std::time::Instant::now();
    for (i, (left_img_path, right_img_path)) in
        camera_images[0].iter().zip(&camera_images[1]).enumerate()
    {
        let left_image = image::open(left_img_path)?.to_luma8();
        let right_image = image::open(right_img_path)?.to_luma8();
        estimator.process_frame(&left_image, &right_image)?;
        let [curr_points0, curr_points1] = &estimator.tracker.get_track_points();

        if let Some(rec) = &rec {
            let timestamp_ns = left_img_path
                .file_stem()
                .and_then(|s| s.to_str())
                .and_then(|s| s.parse::<i64>().ok())
                .unwrap_or(0);
            if timestamp_ns == 0 {
                log::warn!(
                    "Failed to parse timestamp from filename '{}'",
                    left_img_path.display()
                );
                rec.set_time_sequence("frame", i as i64);
            } else {
                rec.set_time(
                    "epoch_time",
                    rerun::TimeCell::from_timestamp_nanos_since_epoch(timestamp_ns),
                );
            }
            log_rerun_image(&rec, &left_image, "camera/left");
            log_rerun_image(&rec, &right_image, "camera/right");
            log_image_keypoints(&rec, curr_points0, &prev_points0, "camera/left");
            log_image_keypoints(&rec, curr_points1, &prev_points1, "camera/right");

            log_active_landmarks(
                &rec,
                &estimator.landmarks,
                &estimator.current_t_w_cam0,
                "/active_landmarks",
                50.0,
            );

            log_pose(
                &rec,
                &estimator.current_t_w_cam0,
                "/current_pose/left",
                0.4,
                camera_frustum.clone(),
            );
            log_pose(
                &rec,
                &(estimator.current_t_w_cam0 * t_cam1_cam0),
                "/current_pose/right",
                0.4,
                camera_frustum.clone(),
            );
            if estimator.new_keyframe_added {
                for (i, kf) in estimator.keyframe_window.keyframes.iter().enumerate() {
                    log_pose(
                        &rec,
                        &kf.t_cam0_w.inverse(),
                        &format!("/keyframe_poses/keyframe_{}", i),
                        0.3,
                        None,
                    );
                }
                log_old_landmarks(&rec, &estimator.removed_good_landmarks, "/old_landmarks");
                log_trajectory(&rec, &estimator.keyframe_trajectory, "/trajectory");
            }
        }

        // for visualization purposes
        prev_points0 = curr_points0.clone();
        prev_points1 = curr_points1.clone();
    }

    let elapsed = start_time.elapsed();
    println!(
        "Processing time: {:.2?}ms average per frame",
        elapsed.as_millis() as f64 / camera_images[0].len() as f64
    );

    Ok(())
}

File: examples/run_stereo.py
"""
Stereo Visual Odometry example using toy_vo Python bindings.
Equivalent to examples/run_stereo.rs with full rerun visualization.
"""

import json
import glob
import time
from pathlib import Path
import argparse

import numpy as np
from numpy.typing import NDArray
import cv2
import rerun as rr
from toy_vo import StereoEstimator, EstimatorParameters


def id_to_color(feature_id: int) -> list[int]:
    """Generate a deterministic color from a feature ID."""
    import hashlib

    h = hashlib.md5(feature_id.to_bytes(8, "little")).digest()
    return [h[0], h[1], h[2]]


def load_camera_config(config_path: str):
    """Load camera model from JSON config file.

    Returns camera_matrix (3x3), distortion_params, model_name, width, height.
    """
    with open(config_path, "r") as f:
        data = json.load(f)

    model_name = list(data.keys())[0]
    params = data[model_name]

    fx = params["fx"]
    fy = params["fy"]
    cx = params["cx"]
    cy = params["cy"]
    width = params["width"]
    height = params["height"]

    camera_matrix = np.array([[fx, 0.0, cx], [0.0, fy, cy], [0.0, 0.0, 1.0]])

    if model_name in ("EUCM", "eucm"):
        distortion_params = np.array([params["alpha"], params["beta"]])
    elif model_name in ("UCM", "ucm"):
        distortion_params = np.array([params["alpha"]])
    elif model_name in ("OpenCVModel5", "opencv5", "OPENCV5"):
        distortion_params = np.array(
            [params["k1"], params["k2"], params["p1"], params["p2"], params["k3"]]
        )
    elif model_name in ("KannalaBrandt4", "kb4", "KB4"):
        distortion_params = np.array(
            [params["k1"], params["k2"], params["k3"], params["k4"]]
        )
    elif model_name in ("FovCamera", "fov_camera", "FOV_CAMERA"):
        distortion_params = np.array([params["w"]])
    else:
        raise ValueError(f"Unknown camera model: {model_name}")

    return camera_matrix, distortion_params, model_name, width, height


def load_extrinsics(extrinsics_path: str):
    """Load stereo extrinsics (T_cam1_cam0) from JSON.

    Returns rvec (3,) and tvec (3,) numpy arrays.
    """
    with open(extrinsics_path, "r") as f:
        data = json.load(f)

    rtvec = data["rtvecs"][1]
    rvec = np.array(rtvec["rvec"])
    tvec = np.array(rtvec["tvec"])
    return rvec, tvec


def load_camera_images(root: str, cam_idx: int) -> list[str]:
    """Load sorted list of image paths for a given camera."""
    pattern = str(Path(root) / f"mav0/cam{cam_idx}/data/*.png")
    paths = sorted(glob.glob(pattern))
    if not paths:
        pattern = str(Path(root) / f"mav0/cam{cam_idx}/data/*.jpg")
        paths = sorted(glob.glob(pattern))
        if not paths:
            raise ValueError(f"No images found for camera {cam_idx} in {root}")
    return paths


def rvec_tvec_to_4x4(rvec: NDArray, tvec: NDArray) -> NDArray[np.float64]:
    """Convert axis-angle rotation vector and translation to 4x4 homogeneous matrix."""
    from scipy.spatial.transform import Rotation

    rot = Rotation.from_rotvec(rvec)
    mat = np.eye(4)
    mat[:3, :3] = rot.as_matrix()
    mat[:3, 3] = tvec
    return mat


def pose_to_rerun_transform(pose_4x4: NDArray) -> "rr.Transform3D":
    """Convert a 4x4 pose matrix to a rerun Transform3D."""
    from scipy.spatial.transform import Rotation

    rot_mat = pose_4x4[:3, :3]
    quat = Rotation.from_matrix(rot_mat).as_quat()  # [x, y, z, w]
    translation = pose_4x4[:3, 3]

    return rr.Transform3D.from_translation_rotation(
        translation=translation.tolist(),
        rotation=rr.Quaternion(xyzw=quat.tolist()),
    )


def log_rerun_image(image: NDArray, entity_path: str):
    """Log a grayscale image to rerun as encoded JPEG."""
    _, buf = cv2.imencode(
        ".jpg", image
    )  # Ensure OpenCV is used to encode (for consistency)
    rr.log(
        entity_path, rr.EncodedImage(contents=buf.tobytes(), media_type="image/jpeg")
    )


def log_image_keypoints(
    curr_keypoints: dict[int, tuple[float, float]],
    prev_keypoints: dict[int, tuple[float, float]],
    image_entity_path: str,
):
    """Log 2D keypoints and optical flow lines on an image."""

    # Points with per-ID colors
    points = []
    colors = []
    for fid, (x, y) in curr_keypoints.items():
        points.append([x + 0.5, y + 0.5])
        colors.append(id_to_color(fid))

    if points:
        rr.log(
            f"{image_entity_path}/points",
            rr.Points2D(points, colors=colors),
        )

    # Lines from previous to current keypoints
    line_strips = []
    line_colors = []
    for fid, (curr_x, curr_y) in curr_keypoints.items():
        if fid in prev_keypoints:
            prev_x, prev_y = prev_keypoints[fid]
            line_strips.append(
                [[prev_x + 0.5, prev_y + 0.5], [curr_x + 0.5, curr_y + 0.5]]
            )
            line_colors.append(id_to_color(fid))

    if line_strips:
        rr.log(
            f"{image_entity_path}/lines",
            rr.LineStrips2D(line_strips, colors=line_colors),
        )


def log_pose(
    pose_4x4: NDArray,
    entity_path: str,
    size: float,
    camera_frustum=None,
):
    """Log a 3D pose (transform + axes or camera frustum)."""

    rr.log(
        entity_path,
        rr.Transform3D(translation=pose_4x4[:3, 3], mat3x3=pose_4x4[:3, :3]),
    )
    if camera_frustum is not None:
        rr.log(entity_path, camera_frustum)
    else:
        rr.log(entity_path, rr.TransformAxes3D(axis_length=size))


def log_active_landmarks(
    landmarks_with_ids: dict[int, tuple[float, float, float]],
    current_pose_4x4: NDArray,
    entity_path: str,
    range_threshold: float,
):
    """Log active 3D landmarks, filtered by distance from current camera."""

    cam_position = current_pose_4x4[:3, 3]
    range_threshold2 = range_threshold * range_threshold

    points = []
    colors = []
    for fid, (x, y, z) in landmarks_with_ids.items():
        pt = np.array([x, y, z])
        dist2 = float(np.sum((pt - cam_position) ** 2))
        if dist2 < range_threshold2:
            points.append([x, y, z])
            colors.append(id_to_color(fid))

    if points:
        rr.log(entity_path, rr.Points3D(points, colors=colors))


def log_old_landmarks(
    landmarks: NDArray,
    entity_path: str,
):
    """Log removed landmarks as 3D points."""

    if landmarks.shape[0] == 0:
        return
    rr.log(entity_path, rr.Points3D(landmarks.tolist()))


def log_trajectory(
    trajectory: NDArray,
    entity_path: str,
):
    """Log the keyframe trajectory as a 3D line strip.

    Trajectory rows are flattened 4x4 T_cam0_w matrices.
    We invert each to get world position.
    """

    if trajectory.shape[0] == 0:
        return

    points = []
    for i in range(trajectory.shape[0]):
        t_cam0_w = trajectory[i].reshape(4, 4)
        # Invert to get T_w_cam0, then extract translation
        t_w_cam0 = np.linalg.inv(t_cam0_w)
        points.append(t_w_cam0[:3, 3].tolist())

    if points:
        rr.log(entity_path, rr.LineStrips3D([points]))


def make_camera_frustum(camera_matrix: NDArray, width: int, height: int):
    """Create a rerun Pinhole from camera intrinsics."""
    import rerun as rr

    fx = camera_matrix[0, 0]
    fy = camera_matrix[1, 1]
    cx = camera_matrix[0, 2]
    cy = camera_matrix[1, 2]

    return rr.Pinhole(
        focal_length=[fx, fy],
        resolution=[width, height],
        principal_point=[cx, cy],
        image_plane_distance=0.1,
    )


def main():

    parser = argparse.ArgumentParser(description="Stereo Visual Odometry")
    parser.add_argument(
        "-d", "--dataset-folder", required=True, help="Path to dataset folder"
    )
    parser.add_argument(
        "-c", "--config-folder", required=True, help="Path to config folder"
    )
    parser.add_argument(
        "--rerun", action="store_true", help="Enable rerun visualization"
    )
    args = parser.parse_args()

    # Load camera configurations
    cam0_matrix, cam0_dist, cam0_model, cam0_w, cam0_h = load_camera_config(
        f"{args.config_folder}/cam0.json"
    )
    cam1_matrix, cam1_dist, cam1_model, cam1_w, cam1_h = load_camera_config(
        f"{args.config_folder}/cam1.json"
    )

    # Load extrinsics
    rvec, tvec = load_extrinsics(f"{args.config_folder}/extrinsics.json")
    t_cam0_cam1_4x4 = np.linalg.inv(rvec_tvec_to_4x4(rvec, tvec))

    # Create estimator
    estimator = StereoEstimator(
        cam0_camera_matrix=cam0_matrix,
        cam0_distortion_params=cam0_dist,
        cam0_model_name=cam0_model,
        cam0_width=cam0_w,
        cam0_height=cam0_h,
        cam1_camera_matrix=cam1_matrix,
        cam1_distortion_params=cam1_dist,
        cam1_model_name=cam1_model,
        cam1_width=cam1_w,
        cam1_height=cam1_h,
        t_cam1_cam0_rvec=rvec,
        t_cam1_cam0_tvec=tvec,
        params=EstimatorParameters(
            tracker_optical_flow_levels=8,
            tracker_grid_size=30,
            translation_threshold=2.0,
        ),
    )

    # Setup rerun
    camera_frustum = None
    if args.rerun:
        import rerun as rr

        rr.init("vo")
        rr.spawn(port=9875)
        camera_frustum = make_camera_frustum(cam0_matrix, cam0_w, cam0_h)
        rr.log("/", rr.ViewCoordinates.RDF, static=True)

    # Load images
    left_images = load_camera_images(args.dataset_folder, 0)
    right_images = load_camera_images(args.dataset_folder, 1)
    assert len(left_images) == len(right_images) and len(left_images) > 0, (
        "Mismatched or empty camera images"
    )
    print(f"Found {len(left_images)} stereo image pairs")

    prev_points0: dict[int, tuple[float, float]] = {}
    prev_points1: dict[int, tuple[float, float]] = {}

    start_time = time.time()
    for i, (left_path, right_path) in enumerate(zip(left_images, right_images)):
        left_img = cv2.imread(left_path, cv2.IMREAD_GRAYSCALE)
        right_img = cv2.imread(right_path, cv2.IMREAD_GRAYSCALE)

        estimator.process_frame(left_img, right_img)

        curr_points = estimator.track_points
        curr_points0 = curr_points["cam0"]
        curr_points1 = curr_points["cam1"]

        if args.rerun:
            # Set timeline
            timestamp_ns = 0
            try:
                stem = Path(left_path).stem
                timestamp_ns = int(stem)
            except ValueError:
                pass
            if (
                timestamp_ns < 1262304000000000000
            ):  # If timestamp is before 2010, assume it's a frame index instead
                rr.set_time("frame", sequence=i)
            else:
                rr.set_time("epoch_time", timestamp=timestamp_ns / 1e9)

            # Log images
            log_rerun_image(left_img, "camera/left")
            log_rerun_image(right_img, "camera/right")

            # Log keypoints and optical flow
            log_image_keypoints(curr_points0, prev_points0, "camera/left")
            log_image_keypoints(curr_points1, prev_points1, "camera/right")

            # Log active landmarks
            log_active_landmarks(
                estimator.landmarks_with_ids,
                estimator.current_pose,
                "/active_landmarks",
                50.0,
            )

            # Log current pose (left)
            log_pose(
                estimator.current_pose,
                "/current_pose/left",
                0.4,
                camera_frustum,
            )

            # Log current pose (right) = T_w_cam0 @ T_cam1_cam0
            t_w_cam1 = estimator.current_pose @ t_cam0_cam1_4x4
            log_pose(
                t_w_cam1,
                "/current_pose/right",
                0.4,
                camera_frustum,
            )

            # On new keyframe: log keyframe poses, old landmarks, trajectory
            if estimator.new_keyframe_added:
                keyframe_poses = estimator.keyframe_poses
                for ki in range(keyframe_poses.shape[0]):
                    kf_pose = keyframe_poses[ki].reshape(4, 4)
                    log_pose(
                        kf_pose,
                        f"/keyframe_poses/keyframe_{ki}",
                        0.3,
                        None,
                    )

                log_old_landmarks(estimator.removed_good_landmarks, "/old_landmarks")
                log_trajectory(estimator.keyframe_trajectory, "/trajectory")

        prev_points0 = curr_points0
        prev_points1 = curr_points1

    elapsed = time.time() - start_time
    avg_ms = (elapsed * 1000) / len(left_images)
    print(f"Processing time: {avg_ms:.2f}ms average per frame")


if __name__ == "__main__":
    main()
